Wikipedia/Code/WMFLanguagesViewController.m

```diff
@@ -433,6 +449,31 @@ - (void)languagesController:(WMFLanguagesViewController*)controller didSelectLan
     [self.delegate languagesController:self didUpdatePreferredLanguages:[MWKLanguageLinkController sharedInstance].preferredLanguages];
 }
 
+- (BOOL)shouldShowFooterForSection:(NSInteger)section {
+    return (self.showPreferredLanguages && (section == 0));
+}
+
+- (nullable UIView*)tableView:(UITableView*)tableView viewForFooterInSection:(NSInteger)section; {
+    if ([self shouldShowFooterForSection:section]) {
+        WMFArticleLanguagesSectionFooter* footer = (id)[tableView dequeueReusableHeaderFooterViewWithIdentifier:[WMFArticleLanguagesSectionFooter wmf_nibName]];
+        footer.title = MWLocalizedString(@"settings-primary-language-details", nil);
+        return footer;
+    } else {
+        return nil;
+    }
+}
+
+- (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath {
+    WMFLanguageCell* cell = (WMFLanguageCell*)[super tableView:tableView cellForRowAtIndexPath:indexPath];
+    cell.isPrimary = (indexPath.row == 0) ? YES : NO;
+    return cell;
+}
+
+- (void)tableView:(UITableView*)tableView moveRowAtIndexPath:(NSIndexPath*)sourceIndexPath toIndexPath:(NSIndexPath*)destinationIndexPath {
+    [super tableView:tableView moveRowAtIndexPath:sourceIndexPath toIndexPath:destinationIndexPath];
+    [tableView reloadData];
```

Comment:
Should use insertion APIs instead of reloading.

usually the only reason to reload a table is if the entire dataset is changing or if the view is re-appearing.